platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java

```diff
@@ -537,31 +501,88 @@ private void doResolve(
-            resolutionStrategy.getComponentSelection(),
-            resolutionStrategy.isDependencyVerificationEnabled(),
-            resolutionStrategy.getCachePolicy(),
+            componentMetadataProcessorFactory,
+            resolveContext.getComponentSelectionRules(),
+            params.isDependencyVerificationEnabled(),
+            resolveContext.getCachePolicy(),
             // We should not need to know _what_ we're resolving in order to construct a resolver for a set of repositories.
             // The request attributes and schema are used to support filtering components by attributes when using dynamic versions.
             // We should consider just removing that feature and making dynamic version selection dumber.
-            requestAttributes,
-            consumerSchema
+            params.getRootVariant().getAttributes(),
+            params.getRootComponent().getMetadata().getAttributesSchema()
         ));
 
         return new ComponentResolversChain(resolvers);
     }
+
+    private ImmutableList<ResolutionParameters.ModuleVersionLock> getAllVersionLocks(ResolutionParameters params) {
+        if (!params.isDependencyLockingEnabled()) {
+            return params.getModuleVersionLocks();
+        }
+
+        if (params.isFailingOnDynamicVersions()) {
+            throw new InvalidUserCodeException(
+                "Both dependency locking and fail on dynamic versions are enabled. You must choose between the two modes."
+            );
+        } else if (params.isFailingOnChangingVersions()) {
+            throw new InvalidUserCodeException(
+                "Both dependency locking and fail on changing versions are enabled. You must choose between the two modes."
+            );
+        }
+
+        return ImmutableList.<ResolutionParameters.ModuleVersionLock>builder()
+            .addAll(getLockfileLocks(params))
+            .addAll(params.getModuleVersionLocks())
+            .build();
+    }
+
+    private ImmutableList<ResolutionParameters.ModuleVersionLock> getLockfileLocks(ResolutionParameters params) {
+        DependencyLockingState dependencyLockingState = dependencyLockingProvider.loadLockState(
+            params.getDependencyLockingId(),
+            params.getResolutionHost().displayName()
+        );
+
+        boolean strict = dependencyLockingState.mustValidateLockState();
+
+        Set<ModuleComponentIdentifier> lockedDependencies = dependencyLockingState.getLockedDependencies();
+        ImmutableList.Builder<ResolutionParameters.ModuleVersionLock> locks = ImmutableList.builderWithExpectedSize(lockedDependencies.size());
+        for (ModuleComponentIdentifier lockedDependency : lockedDependencies) {
+            locks.add(new ResolutionParameters.ModuleVersionLock(
+                lockedDependency.getModuleIdentifier(),
+                lockedDependency.getVersion(),
+                "Dependency version enforced by Dependency Locking",
```

Comment:
❓ Is this change necessary?  This feels like a step backwards from the previous message (i.e. "dependency was locked to version ${lockVersion}..."). This new message may not be meaningful to users.  At least including the version makes it clear what value is in effect.  Maybe I'm not understanding, but don't we have everything we need here to construct a similar message to what we had before?

Reply:
I agree this is in a way a step backwards -- but I think it is quite a minor one -- the reason being that the version is always very close by. Consider the output of the `dependencyInsight` task where this reason is visible (on master -- without this change):

```
> Task :dependencyInsight
org:foo:1.0
  Variant default:
    | Attribute Name    | Provided | Requested |
    |-------------------|----------|-----------|
    | org.gradle.status | release  |           |
   Selection reasons:
      - By constraint: dependency was locked to version '1.0'
```

The version is just a few lines above the printed reason.

Same goes for this error message about dependency constraints that cannot be satisfied -- the version is present on the same line:

```
      > Cannot find a version of 'org:foo' that satisfies the version constraints:
           Dependency path ':depLock:unspecified' --> 'org:foo:1.+'
           Dependency path ':depLock:unspecified' --> 'org:foo:{strictly 1.1}'
           Constraint path ':depLock:unspecified' --> 'org:foo:{strictly 1.0}' because of the following reason: dependency was locked to version '1.0'
```

Finally, the version is readily accessible programmatically, right next to the reason, in `ResolvedComponentResult`. You get the reason with `getSelectionReason`, and you get the version from the `getId` method of a `ComponentResult` (the parent interface). 

Now, _why_ did I did I make this change? Its not really related to this PR other than I touched the code and felt I should make the change -- but this causes us to allocate a new String for each component of each dependency graph when dependency locking is enabled. And, these strings are retained in memory for the entire build during sync.

Say we have a 1000 project build, each with ~200 configurations, lets say conservatively 30 resolvable configurations with 10 dependencies each (thats a really small graph!) At 36 bytes per string (at minimum), thats 10MB. Scale up to an average of 100 components per graph and 5000 projects like in some of the larger builds we're seeing and thats 500MB of Strings. 

I recall seeing this at some point when I was analyzing heap dumps.

Reply:
The memory concerns seem compelling, and I see what you mean about the version being nearby.  It does mean the message is sort of useless outside of the rest of the info, but that's a very minor nitpick.  

I do feel like we could improve the message, but I don't have any ideas that don't seem similarly vague:
- "the dependency was locked to this version"
- "the dependency version is locked"
- "a dependency lock enforces this version"

Just thinking out loud, I wonder if we could address the memory issue with the original message by using a constant format value and only creating the string when it's actually rendered.  i.e. `dependency was locked to version '%s'` and `getReason()` calls `String.format()` somehow.  Just sharing the thought - not suggesting we need to run off and investigate this.